Addresses issue #2604: adds an identifier to pages.

tzimisces opened the pull request:
ADD: Extends the header of a Zim-page with an extra field named
'Page-Identifier'. The field gets added to a page on page
creation, and also - if the field is not present yet - when
it is saved.

tzimisces committed:
Addresses issue #2604: adds an identifier to pages.

ADD: Extends the header of a Zim-page with an extra field named
     'Page-Identifier'. The field gets added to a page on page
     creation, and also - if the field is not present yet - when
     it is saved.

tzimisces committed:
Uncommented a test.

The assert was disabled when figuring out why it was failing.

tzimisces committed:
Typo in failed correction of a comment. Classic.

jaap-karssenberg commented:
Initial patch looks good to me; thanks!

Next step would be to add some API around it for plugins to work with:
- Add the page ID to the pages index table & add a method to the index view to retrieve pages by ID
- Add a "getter" or property to the Page object to get the page ID and trigger initialization if the page did not have one yet (for existing notebooks)

Kind regards,

Jaap

tzimisces committed:
Storing and retrieving pages using a unique page identifier.

ADD: Store unique page identifier in database.
ADD: Methods allowing client code to work with page identifiers.

A unique page identifier is added to a page when it is saved.
If an identifier is present in the database, that one will be used.
If an identifier is not present, it will be generated. The existing
signal-wiring will yield in the new identifier being stored in the
database.

tzimisces committed:
FIX: Pages in the cache could go without an ID if they initially did not
have an ID when retrieved via Notebook.get_page.

tzimisces committed:
MOD: When generating a unique iD, set the property also.

tzimisces commented:
Hello Jaap,

Sorry for the late reply, it took a while for me to make some time to look into this further.

I've fabricated something which works for new and existing pages. New pages and pages without unique ID's get one assigned when a page is saved.
I'm assuming that this always happens via the Notebook so the triggering of signals will cause the ID being picked up and stored in the database.

Editing with an external editor does not work yet (hmm, maybe my assumption made above isn't quite right), and I haven't looked at importing existing files for now - I'll do that later on.

In the mean time, I'm open to suggestions :)

Regards,

  Bob

tzimisces committed:
Circumvent a crash if duplicate page identifiers exist when a page is
moved.

FIX: When moving a page without updating links to it which exist on
     other pages, an exception can occur. This happens when the original
     record has a non-NULL value for the page_identifier field.
     Reason: the original record is kept because of the existing
     link(s), and duplicate page identifiers are not allowed.
     For now it is assumed that the Page-Identifier field in a physical
     page is leading, and that it makes no sense to have a page
     identifier for a non-existing page.

Osndok commented:
It appear that the number in the title did not link the issue, so adding that here: #2604

jaap-karssenberg commented:
Thanks for the patch so far - looks good and I'm tempted to go ahead and merge it. Two items I would like to change/see changed:

1. For the new methods added in the index view and in the page object, I miss test cases
2. The argument for the "page-changed" signal changed from "content" to "tree" - I would prefer to provide the content and the meta data as separate field - current the tree object is mutable so I worry about side effects of using the tree here -- OTOH I do want to make the tree object immutable in the future, so that would make this a lesser issue -- have you changed all places where this signal is used?

tzimisces committed:
Merge branch 'develop' into page-identifier.

tzimisces committed:
Superfluous assignment.

tzimisces committed:
When parsing text, allow for interpreting it to be from a file.

A text buffer could have been read from a file, which implies that it
could hold metadata.

tzimisces committed:
Obviously, text taken from a file, is from a file :)

tzimisces committed:
Added some tests which check the behaviour of importing files into Zim,
in relation to the use of page identifiers.

tzimisces committed:
Fix setting the page identifier on a Page.

The change should be propagated to the metadata.

tzimisces committed:
Updated and added tests.

Also testing new methods on Notebook and Page.

tzimisces committed:
Updated files ignored by Git.

tzimisces committed:
Added argument needs to be present in other related classes.

tzimisces committed:
Fix unnecessary store(). And a typo.

tzimisces commented:
So a finally made some time to get to grips with this - Python isn't a daily routine for me :)

Ad 1) I added a number of testcases for the added / changed functionality.
Ad 2) I merely changed the parameter name to reflect the naming used in the calling method - the object type wasn't changed :) Unless I'm missing something in the semantics, I suppose this is no problem?

While adding the testcases I found some minor misbehaviour - that has been fixed.

The behaviour that when importing a file containing a page identifier, an already registed page with the same page identifier will have its identifier removed, may be not the nicest way of doing things, but then this can be considered a corner case anyway.

tzimisces committed:
Merge branch 'develop' into page-identifier.

Osndok commented on zim/notebook/index/pages.py:
```diff
@@ -76,7 +77,8 @@ def __init__(self, db, layout, filesindexer):
 				mtime TIMESTAMP,
 
 				source_file INTEGER REFERENCES files(id),
-				is_link_placeholder BOOLEAN DEFAULT 0
+				is_link_placeholder BOOLEAN DEFAULT 0,
+				page_identifier TEXT UNIQUE,
```
Do we need a database migration (or version number increment) somewhere to properly support this, or is this change enough for that?

tzimisces replied:
Ah, yes, we do indeed need a database migration. I'll increase the database version to 0.9 - if I'm correct, that triggers a re-initialisation and is the way to go?

Osndok commented on zim/notebook/index/pages.py:
```diff
@@ -200,9 +202,9 @@ def _insert_page(self, pagename, is_link_placeholder, file_id=None):
-				(pagename.name, lowerbasename, sortkey, parent_row['id'], is_link_placeholder, file_id)
+				'INSERT INTO pages(name, lowerbasename, sortkey, parent, is_link_placeholder, source_file, page_identifier)'
+				'VALUES (?, ?, ?, ?, ?, ?, ?)',
+				(pagename.name, lowerbasename, sortkey, parent_row['id'], is_link_placeholder, file_id, None)
```
With just a cursory reading, I'm not sure why we skip the `page_identifier` here.

Osndok commented on zim/notebook/index/pages.py:
```diff
@@ -260,14 +262,41 @@ def update_parent(self, parentname, allow_cleanup=lambda r: True, oldrow=None):
+			# filesystem. This happens when a page is moved, and some other page
+			# refers to it, but its link is not updated.
+			#
+			# TODO Handle duplicate page identifiers on another level.
+			self.db.execute(
+				'UPDATE pages SET mtime=?, page_identifier=NULL WHERE page_identifier=?',
+				(mtime, page_identifier),
+			)
```
I might be reading this wrong, but this smells a bit paranoid, or maybe that we may be trying to solve a problem we don't have in the wrong place.

i.e. I didn't see a migration/schema change, so I don't even know if the db even requires `page_identifier` to be unique, but much like your comment here... it should be handled elsewhere. Having dupes in the database or throwing here is not as bad as silently changing (thus dereferencing) another page when we are unsure which one is 'correct', and it also makes every `update_page` need two db operations when the expected case is that this one will only find "ourself".

I do understand that page_identifier is conceptually unique, but let me see if understand a plausible cause and effect:
1. Given that `PageA` exists with id `1`
2. Some other process duplicates it (thus duplicating it's identifier). e.g. the user does `cp PageA.txt PageB.txt`
3. Zim detects the new `PageB` and imports it
4. When `PageB` is updated, this code wants to remove `PageA`'s identifier, and declare `PageB` as true holder of id `1`
5. When `PageA` is reloaded from disk later, and updated, this code will remove `PageB`'s identifier in the db, and declare that `PageA` is the true holder of id `1`

...do I have that about right? It seems a bit thrashy, but maybe you had a `mv` in mind for step-2.

Assuming we do want to enforce uniqueness (which I could see the value of), then it seems to me that the existing db entry would win (assuming the page is still present at that path), and the newbie that we are updating needs to be assigned a new id, not the other way around.

Osndok reviewed (commented):
Thank you for the work, I'm really looking forward to getting something like this merged/released.
